File: Languages/Dart/Basics/1. Variables & Data Types/2. Data Types.dart
  ////////////////
 // Data Types //
////////////////
// All dart variabels are objects, instances of classes that are subtypes of the object class
// This file describes most of the built in classes in Dart (can be used without importing anything or defining a new class)
// We can check the run time type of a variable using the runtimeType property
// Runtime type may be different from pre compile type as for example unless specified otherwise Maps have runtime type Map<dynamic,dynamic>
void displayType(variable){print(variable.runtimeType);}
Object a = 1;



// Booleans
bool t  = true;
bool f = false;



// Numbers
num b = 1;          // num is a superclass of int and double and has math operators like +, /, abs(), floor() etc. but no bitwise operators like >>>
int c = 1;          // <64 bit integer (64 bits on native platforms and 53 bits for integer part on the web)
double d = 1;       // exactly 64 bit float
var hex = 0xEF;           // use hex code to define the the integer 239
var y = 1.1;              // this is a double as it is a decimal number
var exponents = 1.42e5;   // same as 1.42 x 10^5



// Strings
// Uses UTF-16 encoding
var s1 = 'Single quote can have \'s and "s';
var s2 = "Single quote can have 's and \"s";
var s3 = 'Can use string interpolation like this $s1 if the expression is a single identifier';
var s4 = 'Otherwise us ${s1.toUpperCase()}';
var s5 = '''You can create
          multi-line strings like this''';
var s6 = r'In a raw string, not even \n gets special treatment.';
var s7 = 'Unicode character can be inserted like \u2665 if there hex code is 4 digit and \u{3} if it is > or < 4 digits';



// Collections
// Lists
List arr = [1, 2, 3];

// Sets (exactly like sets in maths or python)
Set halogens = {'fluorine', 'chlorine', 'bromine', 'iodine', 'astatine'};
// create an empty set
Set set1 = {};
var set2 = <String>{};
var set3 = Set();

// Maps  (darts implementation of Hasmaps similar to dictoinaries in python)
var map = {};                             // create an empty set
Map map2 = Map<String, String>();         // 

// Records
// these are immutable and have fixed length like tuples in python but can also have named field as well as positional ones
var record = ('first', a: 2, b: true, 'last');



// Runes
// runes are integers corresponding to the Unicode code point (e.g. ♥ = \u2665 = 9829) that string are made out of
Runes r1 = Runes("sr");         // iterable of runes stores as the integers (115, 114) 
Runes rune = s1.runes;



// Symbols
// A Symbol object represents an operator or identifier declared in a Dart program
// they're invaluable for APIs that refer to identifiers by name, because minification changes identifier names but not identifier symbols
// Symbol literals are compile-time constants.
Symbol symbol = #radix;       // gets the symbol for the radix identifier



// Inline function types
// Function types define the datatypes of function args
// They are used when you store functions like variables 
// e.g. in a list, as a return type, variable declaration, ... 
void Function(String, int)? inlineFunctionTypes(String s1) {                // the retrun type is this function is either null or void Function(String, int)? (a function that takes a string and int)
  final bool Function(Map) f1;                                              // this variables type is bool Function(Map) (a funtion that returns a bool and takes a map as an arg)
  final List<Set Function(int)> f2;                                         // this Lists's elements type is Set Function(int)
  void Function(String, int)? last;                                         // this variable matches the return variable so the function can return it
  return last;
}



// Converting and casting
var S_to_I = int.parse('1');                              // String -> int
var S_to_D = double.parse('1.1');                         // String -> double
var S_to_B = bool.parse('true');                          // String -> bool
var S_to_L = 'list'.split('');                            // String -> List

var I_to_S = 1.toString();                      // int -> String
var I_to_D = 1.toDouble();                      // int -> Double

String piAsString = 3.14159.toStringAsFixed(2);         // double -> String



// Typedefs
// Allows you to create new syntax to refer to a new type 
// they must be defined at the top level of a file (global scope)
typedef IntList = List<int>;
IntList il = [1, 2, 3];

Map<String, List<String>> m1 = {};                    // Verbose
typedef ListMapper<X> = Map<X, List<X>>;
ListMapper<String> m2 = {};                           // Same thing but shorter and clearer.

typedef Compare<T, P> = int Function(T a, P b);       // can use function typedefs but inline function types are preffered
void typedef(){
  final int Function<T, P>(T a, P b) g;                 // same as this
}

File: Languages/C#/Basics/1. Variable & Data Types/2. Data Types.cs
  ////////////////
 // Data Types //
////////////////
// All C# variables are objects, instances of classes that are subtypes of the object class
// This file describes most of the built in classes in Dart (can be used without importing anything or defining a new class)
// We can check the run time type of a variable using the runtimeType property
// Runtime type may be different from pre compile type as for example unless specified otherwise Maps have runtime type Map<dynamic,dynamic>
Object a = 1;

// Booleans
bool t  = true;
bool f = false;


// Numbers
// Each of the following has an unsigned counterpart byte, ushort, uint, ulong, 
sbyte n1 = 8;                   // 8 bit whole number
short n2 = 16;                  // 16 bit whole number
int n3 = 32;                    // 32 bit whole number
long n4 = 64L;                  // 64 bits whole number (use l/L suffix for implicit long typing)
float n5 = 32F;                 // 32 bit floating point number (use f/F suffix for implicit float typing)
double n6 = 64.0D;              // 64 bit floating point number (use d/D or .0 suffix for implicit double typing)
decimal n7 = 128M;              // 64 bit floating point number (use m/M for implicit decimal typing)

var exponents = 1.42e5;//= 1.42 x10^5           // use e/E like x10^ to define floating point numbers (defaults to double unless f/F suffix)
var hex = 0xEF;// = 239                         // use 0x/0X prefix and hexadecimal number to define the integer 
var bin = 0b11;// = 3                           // use 0b/0B prefix and binary number to define the integer 


// Strings and chars
char ch = 'D';              // use '' to define a single unicode character (16 bits or 2 bytes)
string str = "Hello";       // use "" to define a sequence of chars


// Finding types
typeof(a);          // gets runtime type of instance 
a.GetType;          // runtime type also stored as a property


// Casting and conversion
// We cannot change the type of a variable but we can convert or cast it's value to a different type and then save it as a new variable
// Implicit casting converts a smaller type to a larger type size (in terms of bits stored) e.g. int ->  double
// Explicit casting converts a larger type to a smaller type size e.g.  double -> int
// Converting runs some method that creates a new value using the old value somehow
double c2 = 9;                      // implicit cast doesn't change stored bits, just tells the compiler to read them differently
int c1 = (int)42.5;                 // explicit cast needs to drop some bits (.5 in this case) so explicit syntax is required
int c3 = Convert.ToInt32("123");    // conversion parses string into an integer
